testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::tx_emitter::TxStats;
+use std::{
+    fmt,
+    fmt::Debug,
+    ops::Sub,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+const DEFAULT_HISTOGRAM_CAPACITY: usize = 2048;
+const DEFAULT_HISTOGRAM_STEP_MILLIS: u64 = 50;
+
+pub struct AtomicHistogramAccumulator {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<AtomicU64>,
+}
+
+impl Default for AtomicHistogramAccumulator {
+    fn default() -> AtomicHistogramAccumulator {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+    }
+}
+
+impl AtomicHistogramAccumulator {
+    pub fn new(size: usize, step: u64) -> AtomicHistogramAccumulator {
+        let mut buf = Vec::with_capacity(size);
+        for _i in 0..size {
+            buf.push(AtomicU64::new(0));
+        }
+        Self {
+            capacity: size,
+            step_millis: step,
+            buckets: buf,
+        }
+    }
+
+    pub fn snapshot(&self) -> AtomicHistogramSnapshot {
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.capacity {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: buf,
+        }
+    }
+
+    fn get_bucket_num(&self, latency: u64) -> u64 {
+        let bucket_num = latency / self.step_millis;
+        if bucket_num >= self.capacity as u64 - 2 {
+            return self.capacity as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_latency_ms(&self, latency_ms: u64, num_requests: u64) {
+        let bucket_num = self.get_bucket_num(latency_ms);
+        self.buckets[bucket_num as usize].fetch_add(num_requests as u64, Ordering::Relaxed);
+    }
+}
+
+pub struct AtomicHistogramSnapshot {
+    capacity: usize,
+    step_millis: u64,
+    pub(crate) buckets: Vec<u64>,
+}
+
+impl Default for AtomicHistogramSnapshot {
+    fn default() -> AtomicHistogramSnapshot {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+            .snapshot()
+    }
+}
+
+// reserving place for debug print in the future
+impl Debug for AtomicHistogramSnapshot {
```

Comment:
Mm, you can just add `#[dervie(Debug)]` on top of `pub struct AtomicHistogramSnapshot`

Reply:
I tried, it will get error and ask to implement Debug for ```AtomicHistogramSnapshot```

Reply:
Mm, this is hard to believe :)
`#[dervie(Debug)]` is implementing debug for this type, what exactly error message that you saw?

Reply:
```error[E0277]: `atomic_histogram::AtomicHistogramSnapshot` doesn't implement `std::fmt::Debug`
  --> testsuite/cluster-test/src/tx_emitter.rs:77:5
   |
77 |     pub latency_buckets: AtomicHistogramSnapshot,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `atomic_histogram::AtomicHistogramSnapshot` cannot be formatted using `{:?}`
```

Reply:
Did you write `#[dervie(Debug)]` on top of `pub struct AtomicHistogramSnapshot` when you had this error message?
If yes, do you mind pushing version with this to github so we look together

Reply:
sorry for the late reply, I think I figured out that I put it on top of wrong place. ...my bad